File: yjs-worker/src/types/yjs_document_state.ts
export type YjsDocumentUpdate = {
  updateSequence: number;
  payload: Buffer;
};

export type YjsDocumentState = {
  snapshot: Buffer;
  stateVector: Buffer;
  lastUpdateSequence: number;
  compactedUntilSequence: number;
  projectedUntilSequence: number;
  updates: YjsDocumentUpdate[];
};

export function parseYjsDocumentState(payload: Buffer): YjsDocumentState | null {
  if (payload.length < 36) {
    return null;
  }

  const lastUpdateSequence = Number(payload.readBigInt64BE(0));
  const compactedUntilSequence = Number(payload.readBigInt64BE(8));
  const projectedUntilSequence = Number(payload.readBigInt64BE(16));
  const snapshotLength = payload.readUInt32BE(24);
  const stateVectorLength = payload.readUInt32BE(28);
  const updateCount = payload.readUInt32BE(32);
  if (
    !Number.isSafeInteger(lastUpdateSequence) ||
    !Number.isSafeInteger(compactedUntilSequence) ||
    !Number.isSafeInteger(projectedUntilSequence) ||
    lastUpdateSequence < 0 ||
    compactedUntilSequence < 0 ||
    compactedUntilSequence > lastUpdateSequence ||
    projectedUntilSequence < -1 ||
    projectedUntilSequence > lastUpdateSequence
  ) {
    return null;
  }

  let offset = 36;
  if (snapshotLength > payload.length - offset) {
    return null;
  }
  const snapshot = payload.subarray(offset, offset + snapshotLength);
  offset += snapshotLength;

  if (stateVectorLength > payload.length - offset) {
    return null;
  }
  const stateVector = payload.subarray(offset, offset + stateVectorLength);
  offset += stateVectorLength;

  const updates: YjsDocumentState["updates"] = [];
  let expectedUpdateSequence = compactedUntilSequence + 1;
  for (let index = 0; index < updateCount; index += 1) {
    if (payload.length - offset < 12) {
      return null;
    }

    const updateSequence = Number(payload.readBigInt64BE(offset));
    const updateLength = payload.readUInt32BE(offset + 8);
    offset += 12;
    if (
      !Number.isSafeInteger(updateSequence) ||
      updateSequence !== expectedUpdateSequence ||
      updateLength > payload.length - offset
    ) {
      return null;
    }

    updates.push({
      updateSequence,
      payload: payload.subarray(offset, offset + updateLength),
    });
    offset += updateLength;
    expectedUpdateSequence += 1;
  }

  if (offset !== payload.length || expectedUpdateSequence - 1 !== lastUpdateSequence) {
    return null;
  }

  return {
    snapshot,
    stateVector,
    lastUpdateSequence,
    compactedUntilSequence,
    projectedUntilSequence,
    updates,
  };
}

export function parseYjsUpdateSequence(payload: Buffer): number | null {
  if (payload.length !== 8) {
    return null;
  }

  const updateSequence = Number(payload.readBigInt64BE());
  if (!Number.isSafeInteger(updateSequence) || updateSequence < 1) {
    return null;
  }

  return updateSequence;
}

File: app/realtime/types/yjs_document_state.go
package realtimetypes

import (
	"bytes"
	"encoding/binary"
	"errors"
	"math"
)

type YjsDocumentState struct {
	Snapshot               []byte
	StateVector            []byte
	LastUpdateSequence     int64
	CompactedUntilSequence int64
	ProjectedUntilSequence int64
	Updates                []YjsDocumentUpdate
}

type YjsDocumentUpdate struct {
	UpdateSequence int64
	Payload        []byte
}

func (s YjsDocumentState) MarshalBytes() ([]byte, error) {
	if len(s.Snapshot) > math.MaxUint32 || len(s.StateVector) > math.MaxUint32 || len(s.Updates) > math.MaxUint32 {
		return nil, errors.New("invalid yjs document state")
	}
	if s.LastUpdateSequence < 0 ||
		s.CompactedUntilSequence < 0 ||
		s.CompactedUntilSequence > s.LastUpdateSequence ||
		s.ProjectedUntilSequence < -1 ||
		s.ProjectedUntilSequence > s.LastUpdateSequence {
		return nil, errors.New("invalid yjs document state")
	}

	payload := bytes.NewBuffer(make([]byte, 0, 36+len(s.Snapshot)+len(s.StateVector)))
	if err := binary.Write(payload, binary.BigEndian, s.LastUpdateSequence); err != nil {
		return nil, err
	}
	if err := binary.Write(payload, binary.BigEndian, s.CompactedUntilSequence); err != nil {
		return nil, err
	}
	if err := binary.Write(payload, binary.BigEndian, s.ProjectedUntilSequence); err != nil {
		return nil, err
	}
	if err := binary.Write(payload, binary.BigEndian, uint32(len(s.Snapshot))); err != nil {
		return nil, err
	}
	if err := binary.Write(payload, binary.BigEndian, uint32(len(s.StateVector))); err != nil {
		return nil, err
	}
	if err := binary.Write(payload, binary.BigEndian, uint32(len(s.Updates))); err != nil {
		return nil, err
	}
	if _, err := payload.Write(s.Snapshot); err != nil {
		return nil, err
	}
	if _, err := payload.Write(s.StateVector); err != nil {
		return nil, err
	}

	for _, update := range s.Updates {
		if len(update.Payload) > math.MaxUint32 {
			return nil, errors.New("invalid yjs document update")
		}
		if err := binary.Write(payload, binary.BigEndian, update.UpdateSequence); err != nil {
			return nil, err
		}
		if err := binary.Write(payload, binary.BigEndian, uint32(len(update.Payload))); err != nil {
			return nil, err
		}
		if _, err := payload.Write(update.Payload); err != nil {
			return nil, err
		}
	}

	return payload.Bytes(), nil
}

func MarshalYjsUpdateSequence(updateSequence int64) []byte {
	payload := make([]byte, 8)
	binary.BigEndian.PutUint64(payload, uint64(updateSequence))

	return payload
}
